oauthenticator/openshift.py

```diff
@@ -127,13 +149,42 @@ async def authenticate(self, handler, data=None):
         )
 
         resp = await http_client.fetch(req)
-        resp_json = json.loads(resp.body.decode('utf8', 'replace'))
+        ocp_user = json.loads(resp.body.decode('utf8', 'replace'))
 
-        return {
-            'name': resp_json['metadata']['name'],
-            'auth_state': {'access_token': access_token, 'openshift_user': resp_json},
+        username = ocp_user['metadata']['name']
+
+        user_info = {
+            'name': username,
+            'auth_state': {'access_token': access_token, 'openshift_user': ocp_user},
         }
 
+        if self.allowed_groups or self.admin_groups:
+            user_info = await self._add_openshift_group_info(user_info)
+
+        return user_info
+
+    async def _add_openshift_group_info(self, user_info: dict):
+        """
+        Use the group info stored on the OpenShift User object to determine if a user
+        is authenticated based on groups, an admin, or both.
+        """
+        user_groups = set(user_info['auth_state']['openshift_user']['groups'])
+
+        if self.admin_groups:
+            is_admin = self.user_in_groups(user_groups, self.admin_groups)
+
+        user_in_allowed_group = self.user_in_groups(user_groups, self.allowed_groups)
+
+        if self.admin_groups and (is_admin or user_in_allowed_group):
+            user_info['admin'] = is_admin
```

Comment:
@wseaton I did a bit more testing and it seams that the `admin` "flag" is not working properly with this flow:

1. group `odh-admins` is set in `admin_groups`
1. user `vpavlin` is in `odh-admins`
1. `is_admin` is set for `vpavlin`
1 `vpavlin` is removed from `odh-admins`
1. log out and log back in
1. `vpavlin` is *still* `admin` (I think the `is_admin` is set in JH DB at that point)
1. go to `Admin` section and remove `admin` privilege for `vpavlin`
2. `vpavlin` is *no longer* `admin`

I think we might need to set `user_info['admin']` always to make sure that it properly respects the groups configuration

Reply:
Ok, I made a mistake in my flow up there - what I actually did was not to remove a user, but remove the `admin_group` which basically made oauthenticator skip the whole `admin` section.

When I did the test correctly, it worked as expected.